src/RamIndexAnalysis.h

```diff
@@ -34,16 +34,160 @@
 #include <utility>
 #include <vector>
 
-#define RIA_NIL 0
-#define RIA_INF -1
-
 // define if enable unit tests
 #define M_UNIT_TEST
 
 namespace souffle {
 
 class RamTranslationUnit;
 
+/** search signature of a RAM operation; each bit represents an attribute of a relation.
+ * A one represents that the attribute has an assigned value; a zero represents that
+ * no value exists (i.e. attribute is unbounded) in the search. */
+class SearchSignature {
+public:
+    explicit SearchSignature(size_t arity, uint64_t mask) : bits(arity, false) {
+        size_t len = bits.size();
+        for (size_t i = 0; i < len; ++i) {
+            bits[i] = (mask % 2);
+            mask /= 2;
+        }
+    }
```

Comment:
This is effectively `boost::dynamic_bitset` (or something near to it). Could you rename it to `dynamic_bitset`, move it to its own header, and add an alias `using SearchSignature = dynamic_bitset`? It could be handy on its own.

Adding basic iterators support would make this compatible with the many existing helpers (e.g. `join`).

Reply:
I've implemented this. We have moved away from the bitset/dynamic bitset like format to a much nicer vector<AttributeConstraint> where AttributeConstraint is an enum representing the states that Bernhard described above.